src/Mesh/Gradients.C

```diff
@@ -0,0 +1,165 @@
+// *****************************************************************************
+/*!
+  \file      src/Mesh/Gradients.C
+  \copyright 2012-2015, J. Bakosi, 2016-2017, Los Alamos National Security, LLC.
+  \brief     Functions computing gradients on unstructured meshes for tetrahedra
+  \details   Functions computing gradients using linear finite element shape
+             functions on unstructured meshes for tetrahedra.
+*/
+// *****************************************************************************
+
+#include "Exception.h"
+#include "Gradients.h"
+#include "Vector.h"
+
+namespace tk {
+
+std::array< tk::real, 3 >
+nodegrad( std::size_t node,
+          const std::array< std::vector< tk::real >, 3 >& coord,
+          const std::vector< std::size_t >& inpoel,
+          const std::pair< std::vector< std::size_t >,
+                           std::vector< std::size_t > >& esup,
+          const tk::Fields& U,
+          ncomp_t c )
+// *****************************************************************************
+//  Compute gradient at a mesh node
+//! \param[in] node Node id at which to compute gradient
+//! \param[in] coord Mesh node coordinates
+//! \param[in] inpoel Mesh element connectivity
+//! \param[in] esup Linked lists storing elements surrounding points, see
+//!    tk::genEsup()
+//! \param[in] U Field vector whose component gradient to compute
+//! \param[in] c Scalar component to compute gradient of
+//! \return Gradient of U(c) at mesh node
+// *****************************************************************************
+{
+  Assert( c < U.nprop(), "Indexing out of field data" );
+
+  const auto& x = coord[0];
+  const auto& y = coord[1];
+  const auto& z = coord[2];
+
+  // storage for gradient and volume at the mesh node
+  std::array< tk::real, 3 > g{{ 0.0, 0.0, 0.0 }};
+  tk::real vol = 0.0;
+
+  // loop over cells surrounding mesh node
+  for (auto k=esup.second[node]+1; k<=esup.second[node+1]; ++k) {
```

Comment:
Could we make this linked list indexing more transparent? It's not a huge problem, and I don't know how best to do it, but it seems a little obtuse

Reply:
I would also like that. Any suggestion? I cannot think of anything obvious right away.

Reply:
Ah, I think I just came up with a cool idea for this. Thanks for suggesting this! I think one should be able to write a container around `esup` and define iterators, `begin()`, `end()`, etc., so that a range-based `for` loop could be written around it by client code, and the loop variable would return `esup.first[k]`. _That_, in a nice generic fashion, would indeed render a LOT of such code less error-prone a lot nicer looking. I will see what I can do - unless you want to take a crack at it. ;-) Let me know.

Reply:
Here is one way to get started: https://stackoverflow.com/a/31457319.